internal/util/dependency/kv/kv_client_handler.go

```diff
@@ -0,0 +1,92 @@
+package kvfactory
+
+import (
+	"sync"
+
+	"github.com/tikv/client-go/v2/config"
+	"github.com/tikv/client-go/v2/txnkv"
+	clientv3 "go.etcd.io/etcd/client/v3"
+
+	"github.com/milvus-io/milvus/pkg/util/etcd"
+	"github.com/milvus-io/milvus/pkg/util/paramtable"
+)
+
+// ETCDReuser is mainly for sessions. This function allows the etcd client to be made after all the configs
+// have been set and allows for the reuise of the etcd client for a package.
+type ETCDReuser struct {
+	mu       sync.Mutex
+	Client   *clientv3.Client
+	Rootpath string
+}
+
+// Construct an empty reuser
+func NewETCDReuser() *ETCDReuser {
+	return &ETCDReuser{}
+}
+
+// Returns an ETCD client and the metaRootPath, if an error is hit, will panic
+func (reuser *ETCDReuser) GetETCDAndPath() (*clientv3.Client, string) {
+	reuser.mu.Lock()
+	defer reuser.mu.Unlock()
+	// If client/path doesnt exist, create a new one
+	if reuser.Client == nil {
+		var err error
+		reuser.Client, err = NewETCDClient()
+		if err != nil {
+			panic(err)
+		}
+		reuser.Rootpath = GetETCDRootPath()
+	}
```

Comment:
If we move client creation code into NewETCDReuser(), can we get rid of `mu` ?

Reply:
The issue that comes up here is that the configs don't necessarily exist yet or might need to be modified so we cant immediately try to create the client but we rather have to wait for the first call. Im debating having an etcdFactory created at datacoord creation and then pass that into the session_util. Once the rest of the factory changes are done we can swap the etcdFactory with the overarching dependency factory.